velox/dwio/dwrf/common/FileMetadata.h

```diff
@@ -0,0 +1,500 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <optional>
+
+#include "velox/dwio/common/Common.h"
+#include "velox/dwio/common/Options.h"
+#include "velox/dwio/common/exception/Exception.h"
+#include "velox/dwio/dwrf/common/Common.h"
+#include "velox/dwio/dwrf/common/wrap/dwrf-proto-wrapper.h"
+#include "velox/dwio/dwrf/common/wrap/orc-proto-wrapper.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox::dwrf {
+
+/***
+ * Wrapper of StripeInformation Message of proto file
+ * Provides access to attributes of both ORC / DWRF format
+ ***/
+class StripeInformation {
+ public:
+  StripeInformation(
+      uint64_t offset,
+      uint64_t indexLength,
+      uint64_t dataLength,
+      uint64_t footerLength,
+      uint64_t numberOfRows)
+      : offset_(offset),
+        indexLength_(indexLength),
+        dataLength_(dataLength),
+        footerLength_(footerLength),
+        numberOfRows_(numberOfRows) {}
+
+  explicit StripeInformation(const proto::StripeInformation& info);
+
+  explicit StripeInformation(const proto::orc::StripeInformation& info);
+
+  uint64_t offset() const {
+    return offset_;
+  }
+
+  uint64_t length() const {
+    return indexLength_ + dataLength_ + footerLength_;
+  }
+  uint64_t indexLength() const {
+    return indexLength_;
+  }
+
+  uint64_t dataLength() const {
+    return dataLength_;
+  }
+
+  uint64_t footerLength() const {
+    return footerLength_;
+  }
+
+  uint64_t numberOfRows() const {
+    return numberOfRows_;
+  }
+
+  bool hasRawDataSize() const {
+    return rawDataSize_.has_value();
+  }
+
+  uint64_t rawDataSize() const {
+    DWIO_ENSURE(rawDataSize_.has_value());
+    return rawDataSize_.value();
+  }
+
+  bool hasChecksum() const {
+    return checksum_.has_value();
+  }
+
+  int64_t checksum() const {
+    DWIO_ENSURE(checksum_.has_value());
+    return checksum_.value();
+  }
+
+  bool hasGroupSize() const {
+    return groupSize_.has_value();
+  }
+
+  uint64_t groupSize() const {
+    DWIO_ENSURE(groupSize_.has_value());
+    return groupSize_.value();
+  }
+
+  uint32_t keyMetadataSize() const {
+    return keyMetadata_.size();
+  }
+
+  const std::vector<std::string>& keyMetadata() const {
+    return keyMetadata_;
+  }
+
+ private:
+  uint64_t offset_;
+  uint64_t indexLength_;
+  uint64_t dataLength_;
+  uint64_t footerLength_;
+  uint64_t numberOfRows_;
+
+  // DWRF-specific attributes
+  std::optional<uint64_t> rawDataSize_{std::nullopt};
+  std::optional<int64_t> checksum_{std::nullopt};
+  std::optional<uint64_t> groupSize_{std::nullopt};
+  std::vector<std::string> keyMetadata_;
+
+  // ORC-specific attributes
+  std::optional<uint64_t> encryptStripeId_{std::nullopt};
+  std::vector<std::string> encryptedLocalKeys_;
+};
+
+/***
+ * Wrapper of Type Message of proto file
+ * Provides access to attributes of both ORC / DWRF format
+ ***/
+class TypeProto {
+ public:
+  virtual TypeKind kind() const = 0;
```

Comment:
Can we remove the virtual dispatch?  Seems the inheritance is not necessary here.

Reply:
This is needed for hiding template from the user. Many users don't care about if this Type is of DWRF's or ORC's.
Without this, functions like `maxStreamsForType` would need to handle the template argument.